[RenderStyleGen] Generate Style::conservativelyCollectChangedAnimatableProperties

weinig opened the pull request:
#### 799655bac2e5f004a6da1967a7f689947bb70e0d
<pre>
[RenderStyleGen] Generate Style::conservativelyCollectChangedAnimatableProperties
<a href="https://bugs.webkit.org/show_bug.cgi?id=304122">https://bugs.webkit.org/show_bug.cgi?id=304122</a>

Reviewed by Darin Adler.

Generates efficient conservative property change checking by using the
`render-style-storage-path` entries in CSSProperties.json to build a
model of the RenderStyle storage structure using that to build the
recursive walk of the properties.

To make this work, all longhand properties (expect logical, cascade aliases
and select internal properties with no storage) now need to specify a storage
path. For properties where this doesn&apos;t make total sense, like coordinate list
value properties (e.g. background-image, animation-duration, etc), a new kind
of container can be specified, &quot;opaque&quot;, which indicates that the storage path
should only be used for equality checks. That allows us to do things like group
all the properties for a specific coordinated value list, say `background`,
with the path to the background list, `m_nonInheritedData-&gt;backgroundData-&gt;background`,
and then conservatively assert that if that has changed, all the background
properties are potentially changed.

A few new `codegen-properties` entries were added to support this work:

  - `render-style-changed-for-animation-custom`:
    Used for properties that need a custom check in StyleChangedAnimatableProperties.

  - `skip-render-style-getter`, `skip-render-style-initial`, `skip-render-style-setter`:
    A few properties that used to specify `skip-render-style`, like `caret-color`, now
    need to specify a storage path, so they can opt out of specific render style functions.
    Specifying `skip-render-style` sets all of these to true if they are not otherwise set.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/rendering/style/RenderStyleBase.h:
* Source/WebCore/style/StyleChangedAnimatableProperties.cpp:
* Source/WebCore/style/StyleChangedAnimatableProperties.h:
* Source/WebCore/style/StyleChangedAnimatablePropertiesCustom.h: Added.
* Source/WebCore/style/StyleChangedAnimatablePropertiesGenerated.h: Added.
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:

Canonical link: <a href="https://commits.webkit.org/304641@main">https://commits.webkit.org/304641@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9eedd88d4b3f4439da80c75c549ba68df28683d1

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136124 "10 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8480 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47403 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143832 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/137993 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9153 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8325 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104068 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139070 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6648 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/121986 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84909 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/135472 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6323 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3980 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4428 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115602 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40180 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146579 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8164 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40749 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112421 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8180 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/6861 "Found 1 new test failure: fast/webgpu/repro_301290.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112765 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28625 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6242 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118286 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62151 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8212 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36349 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7928 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71771 "Built successfully") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8151 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8004 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSProperties.json:
```diff
@@ -445,11 +449,13 @@
             "codegen-properties": {
                 "style-builder-custom": "Value",
                 "style-extractor-custom": true,
+                "render-style-has-explicitly-set-storage-path": ["m_nonInheritedData", "miscData"],
```
Moves this, but not to alphabetical order as far as I can tell. What order did you intend?

darinadler commented on Source/WebCore/style/StyleChangedAnimatableProperties.h:
```diff
@@ -32,6 +32,10 @@ struct CSSPropertiesBitSet;
 
+// Conservatively answers what CSS properties we should visit for CSS transitions. We do not need
+// to precisely check equivalence before saying "this property needs to be visited". This is tuned
+// based on Speedometer3.0 data.
```
```suggestion
// based on Speedometer 3.0 data.
```
I’m not sure the “is tuned” comment is quite right in terms of verb tense. Maybe more like “was tuned in 2025 based on Speedometer 3.0 data”?

darinadler commented on Source/WebCore/style/StyleChangedAnimatablePropertiesCustom.h:
```diff
@@ -0,0 +1,60 @@
+public:
+    static inline void conservativelyCollectChangedAnimatablePropertiesForCursor(const RenderStyle::InheritedFlags&, const RenderStyle::InheritedFlags&, CSSPropertiesBitSet&);
+    static inline void conservativelyCollectChangedAnimatablePropertiesForZIndex(const StyleBoxData&, const StyleBoxData&, CSSPropertiesBitSet&);
+    static inline void conservativelyCollectChangedAnimatablePropertiesForCaretColor(const StyleRareInheritedData&, const StyleRareInheritedData&, CSSPropertiesBitSet&);
```
When the inline function bodies are in the same source file, not sure it’s valuable to also specify the `inline` in the function declarations.

darinadler commented on Source/WebCore/style/StyleChangedAnimatablePropertiesCustom.h:
```diff
@@ -0,0 +1,60 @@
+namespace WebCore {
+namespace Style {
+
+class ChangedAnimatablePropertiesCustom final {
```
How did you decide on free functions in the `WebCore::Style` namespace vs. functions in a namespace nested inside `WebCore::Style` vs. a class vs. a struct here? These are already such long function names it doesn’t seem to me that they need namespacing.

weinig replied:
Mostly following existing precedent. But I will reconsider.

darinadler commented on Source/WebCore/style/StyleChangedAnimatablePropertiesGenerated.h:
```diff
@@ -0,0 +1,41 @@
+
+namespace Style {
+
+class ChangedAnimatablePropertiesGenerated final {
+public:
+    static void conservativelyCollectChangedAnimatableProperties(const RenderStyle&, const RenderStyle&, CSSPropertiesBitSet&);
+};
```
Why does this need to be a class at all? Why not just a function?

weinig replied:
I was following the pattern of StyleBuilder/StyleExtractor, but I think you are right.

darinadler commented on Source/WebCore/style/StyleDifference.cpp:
```diff
@@ -65,14 +65,14 @@ static bool positionChangeIsMovementOnly(const Style::InsetBox& a, const Style::
 {
     auto nonInheritedDataChangeAffectsVisualOverflow = [&] {
-        if (&a.nonInheritedData() == &b.nonInheritedData())
+        if (a.m_nonInheritedData.ptr() == b.m_nonInheritedData.ptr())
```
Seems like if a data member is intended to be accessed outside the class it’s in, we don’t want to use the `m_` prefix in its name. So I’m surprised to see these changes here. Maybe you have a good style reason for this, related to `friend`?

darinadler reviewed (approved):
This sure does seem like a great idea. I hope there is sufficient test coverage because I certainly wouldn’t be able to spot subtle mistakes from code inspection.

weinig committed:
[RenderStyleGen] Generate Style::conservativelyCollectChangedAnimatableProperties
https://bugs.webkit.org/show_bug.cgi?id=304122

Reviewed by Darin Adler.

Generates efficient conservative property change checking by using the
`render-style-storage-path` entries in CSSProperties.json to build a
model of the RenderStyle storage structure using that to build the
recursive walk of the properties.

To make this work, all longhand properties (expect logical, cascade aliases
and select internal properties with no storage) now need to specify a storage
path. For properties where this doesn't make total sense, like coordinate list
value properties (e.g. background-image, animation-duration, etc), a new kind
of container can be specified, "opaque", which indicates that the storage path
should only be used for equality checks. That allows us to do things like group
all the properties for a specific coordinated value list, say `background`,
with the path to the background list, `m_nonInheritedData->backgroundData->background`,
and then conservatively assert that if that has changed, all the background
properties are potentially changed.

A few new `codegen-properties` entries were added to support this work:

  - `render-style-changed-for-animation-custom`:
    Used for properties that need a custom check in StyleChangedAnimatableProperties.

  - `skip-render-style-getter`, `skip-render-style-initial`, `skip-render-style-setter`:
    A few properties that used to specify `skip-render-style`, like `caret-color`, now
    need to specify a storage path, so they can opt out of specific render style functions.
    Specifying `skip-render-style` sets all of these to true if they are not otherwise set.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/scripts/process-css-properties.py:
* Source/WebCore/rendering/style/RenderStyleBase.h:
* Source/WebCore/style/StyleChangedAnimatableProperties.cpp:
* Source/WebCore/style/StyleChangedAnimatableProperties.h:
* Source/WebCore/style/StyleChangedAnimatablePropertiesCustom.h: Added.
* Source/WebCore/style/StyleChangedAnimatablePropertiesGenerated.h: Added.
* Tools/Scripts/webkitpy/style/checkers/jsonchecker.py:

Canonical link: https://commits.webkit.org/304641@main
